[FLINK-35305][Connector/SQS] Amazon SQS Sink Connector

19priyadhingra opened the pull request:
## Purpose of the change

Implements the Amazon SQS Sink Connector

## Verifying this change

This change added tests and can be verified as follows:

Added unit tests
Added integration tests
Manually verified by running the SQS connector on a local Flink cluster.

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [x] New feature has been introduced
  - If yes, how is this documented? (not applicable / docs / JavaDocs / not documented)

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/archunit-violations/stored.rules:
```diff
@@ -0,0 +1,4 @@
+#
+#Tue Feb 22 12:19:26 CET 2022
+Tests\ inheriting\ from\ AbstractTestBase\ should\ have\ name\ ending\ with\ ITCase=54da9a7d-14d2-4632-a045-1dd8fc665c8f
```
I am not sure if these auto-generated files should be added to PR?

vahmed-hamdy replied:
yes this is part of arch tests build

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsSinkITCase.java:
```diff
@@ -0,0 +1,121 @@
+ */
+@Testcontainers
+@ExtendWith(MiniClusterExtension.class)
+class SqsSinkITCase {
```
This is better suited in `flink-connector-aws-e2e-tests` could we move it?

19priyadhingra replied:
Yes, I can move this but I found all other sinks also has the same tests in their own package so thought of following the same convention

19priyadhingra replied:
Created "flink-connector-aws-sqs-e2e-tests" under "flink-connector-aws-e2e-tests" and move this test there, but I am really struggling with how to run that test there. Running local-stack not helping there and I tried to follow the readme given [here](https://github.com/apache/flink-connector-aws/tree/main/flink-connector-aws-e2e-tests). Downloaded flink-dist-1.19.0.jar and pasted under "flink-connector-aws-e2e-tests" folder and ran "mvn clean verify -Prun-end-to-end-tests -DdistDir=flink-dist-1.19.0.jar" , it is getting failed with error [screenshot attached]. any documentation on how to easily run these test will be super helpful . thanks!

<img width="1671" alt="Screenshot 2024-05-22 at 7 49 54 PM" src="https://github.com/apache/flink-connector-aws/assets/169495197/406cae7a-5eae-421f-8e11-cae88051f78f">

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/testutils/SqsTestUtils.java:
```diff
@@ -0,0 +1,76 @@
+    private static final ObjectMapper MAPPER = createObjectMapper();
+
+    public static SqsClient createSqsClient(String endpoint, SdkHttpClient httpClient) {
+        return AWSServicesTestUtils.createAwsSyncClient(endpoint, httpClient, SqsClient.builder());
```
nit: Is there a reason this is async as well?

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/testutils/SqsTestUtils.java:
```diff
@@ -0,0 +1,76 @@
+                .map(data -> MAPPER.writeValueAsString(ImmutableMap.of("data", data)));
+    }
+
+    public static List<String> getSampleData(int endValue) throws JsonProcessingException {
```
this is unused

19priyadhingra replied:
Removed

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/testutils/SqsTestUtils.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private static void registerModules(ObjectMapper mapper) {
+        mapper.registerModule(new JavaTimeModule())
```
why is this needed?

19priyadhingra replied:
Removed

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsSinkWriterTest.java:
```diff
@@ -0,0 +1,105 @@
+        String testString = "{many hands make light work;";
+        SendMessageBatchRequestEntry record = SendMessageBatchRequestEntry.builder().messageBody(testString).build();
+        assertThat(sinkWriter.getSizeInBytes(record))
+                .isEqualTo(testString.getBytes(StandardCharsets.US_ASCII).length);
```
is this correct? why not `StandardCharsets.UTF_8` ?

19priyadhingra replied:
oops, fixed

vahmed-hamdy commented on flink-connector-aws-base/src/test/java/org/apache/flink/connector/aws/testutils/AWSServicesTestUtils.java:
```diff
@@ -114,6 +116,14 @@ public static void createBucket(S3Client s3Client, String bucketName) {
         }
     }
 
+    public static void createSqs(String sqsName, SqsClient sqsClient) {
```
This shouldn't be in `aws-base` we could move it to `SqsTestUtils`

19priyadhingra replied:
Moved to SqsTestUtils

vahmed-hamdy commented on flink-connector-aws-base/pom.xml:
```diff
@@ -76,6 +76,12 @@ under the License.
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
This should be moved from base

19priyadhingra replied:
removed

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+    private Optional<SendMessageBatchRequestEntry> getFailedRecord(List<SendMessageBatchRequestEntry> requestEntries,
+                                                                   String selectedId)
+    {
+        return requestEntries.stream()
```
Could we avoid java streams as per [coding standards](https://flink.apache.org/how-to-contribute/code-style-and-quality-java/#java-streams)

19priyadhingra replied:
sure. Removed

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequestEntry;
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchResponse;
+
+import java.io.UnsupportedEncodingException;
```
this import is unused
could we please run checkstyle and spotless checks and address the issues?

19priyadhingra replied:
Removed

vahmed-hamdy commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkBuilder.java:
```diff
@@ -0,0 +1,161 @@
+public class SqsSinkBuilder<InputT>
+        extends AsyncSinkBaseBuilder<InputT, SendMessageBatchRequestEntry, SqsSinkBuilder<InputT>> {
+
+    private static final int DEFAULT_MAX_BATCH_SIZE = 10;
```
Is there a reason for choosing such a low value?

19priyadhingra replied:
As per  SendMessageBatch [documentation](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessageBatch.html#:~:text=Because%20the%20batch%20request%20can,xFFFD%20%7C%20%23x10000%20to%20%23x10FFFF), we can at max send 10 messages in a request

hlteoh37 replied:
Do we know if request will be rejected if batch contains > 10 messages? If so, can we add a validation on the builder to ensure customer doesn't set max > 10?

19priyadhingra replied:
Added a validation for 10 max batch size

vahmed-hamdy reviewed (changes requested):
Thanks for the efforts.
Could we remove the ticket link from the PR title, it should be automatically linked using "Autolink"

19priyadhingra commented:
> Thanks for the efforts. Could we remove the ticket link from the PR title, it should be automatically linked using "Autolink"
> 
> We need to add documentation for the new feature as well

1) Updated
2) Sure, will create a documentation on this

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+    /* Flag to whether fatally fail any time we encounter an exception when persisting records */
+    private final boolean failOnError;
+
+    SqsSinkWriter(
```
Removed

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkBuilder.java:
```diff
@@ -0,0 +1,161 @@
+ *
+ * SqsSink<String> sqsSink =
+ *         SqsSink.<String>builder()
+ *                 .setElementConverter(elementConverter)
```
Removed

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        if (response.failed() != null) {
+            LOG.debug("handlePartiallyFailedRequest: SQS Sink failed to write and will retry {} entries to SQS",
```
updated to warn

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsStateSerializer.java:
```diff
@@ -0,0 +1,54 @@
+    {
+        final byte[] requestData = new byte[(int) requestSize];
+        in.read(requestData);
+        return SendMessageBatchRequestEntry.builder().messageBody(new String(requestData, StandardCharsets.UTF_8)).build();
```
yes, test case updated

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsExceptionClassifiersTest.java:
```diff
@@ -0,0 +1,72 @@
+                                AwsErrorDetails.builder().errorCode("NotAuthorized").build())
+                        .build();
+
+        // isFatal returns `true` if an exception is non-fatal
```
I agree with you that it is confusing, but I am not sure why isFatal is implemented like this. I referred [this](https://github.com/apache/flink-connector-aws/blob/c688a8545ac1001c8450e8c9c5fe8bbafa13aeba/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/sink/throwable/AWSExceptionHandler.java#L27) also which says "isFatal returns `true` if an exception is non-fatal"

Also I refer these test cases from existing firehose [test cases](https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/AWSFirehoseExceptionClassifiersTest.java)

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsExceptionClassifiersTest.java:
```diff
@@ -0,0 +1,72 @@
+                        .build();
+
+        // isFatal returns `true` if an exception is non-fatal
+        assertFalse(classifier.isFatal(sqsException, ex -> {}));
```
based on isFatal["isFatal returns true if an exception is non-fatal"] implementation, we have assertFalse here

vahmed-hamdy replied:
Ah I see, got it

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsExceptionClassifiersTest.java:
```diff
@@ -0,0 +1,72 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+/** Unit tests for {@link SqsExceptionClassifiers}. */
+public class SqsExceptionClassifiersTest {
```
add new test case - shouldClassifySocketTimeoutExceptionAsNonFatal

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+    @Override
+    protected long getSizeInBytes(SendMessageBatchRequestEntry requestEntry) {
+        try {
+            return requestEntry.messageBody().getBytes("UTF-8").length;
```
updated

19priyadhingra commented on flink-connector-aws/pom.xml:
```diff
@@ -18,8 +18,8 @@ specific language governing permissions and limitations
 under the License.
 -->
 <project xmlns="http://maven.apache.org/POM/4.0.0"
-		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
reverted

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+            numRecordsOutErrorsCounter.inc(response.failed().size());
+        }
+
+        if (failOnError) {
```
Added

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,272 @@
+                SqsConfigConstants.SQS_CLIENT_USER_AGENT_PREFIX);
+    }
+
+    private static final AWSExceptionHandler SQS_EXCEPTION_HANDLER =
```
Added

19priyadhingra commented:
Thanks for the feedback @vahmed-hamdy. I replied to all the feedback except the documentation one which is still in progress. Is there a package where I suppose to add the documentation for the same?

sap1ens commented:
@19priyadhingra 👋 I've had the opportunity to build and try this connector. Want to share some feedback:

- Currently, it looks like all messages are encoded with base64. It'd be great to be able to not use it, e.g. by setting a custom `ElementConverter`. 
- There are lot of `spotless` violations, you probably want to run `mvn spotless:apply`.

19priyadhingra commented:
> @19priyadhingra 👋 I've had the opportunity to build and try this connector. Want to share some feedback:
> 
> * Currently, it looks like all messages are encoded with base64. It'd be great to be able to not use it, e.g. by setting a custom `ElementConverter`.
> * There are lot of `spotless` violations, you probably want to run `mvn spotless:apply`.

@sam1ens, thanks a lot for the feedback. 
1) Base64 encoding was the one we were using in our production environment and intention for using it was that base64-encoding guarantees that no invalid characters are present in the message, but now I understand that others might not need it. Would you recommend keeping it normal, String?

2) I have a follow up query on spotless violations. How did you run that. Whenever I am trying to do that, it shows build succeeded with no error and I can see "spotless skipped" in build logs[Attached screenshot]. Is there some setting in code which i need to do to enable spotless working for me? 
![image](https://github.com/apache/flink-connector-aws/assets/169495197/20262c6a-9492-476c-a5b2-ce7809664e8a)
Sorry for the basic question, I am working on this package for the first time :)

sap1ens commented:
> Would you recommend keeping it normal, String?

Yes! And if someone needs base64 encoding they can encode it in the `SerializationSchema`.

> I have a follow up query on spotless violations. How did you run that. Whenever I am trying to do that, it shows build succeeded with no error and I can see "spotless skipped" in build logs[Attached screenshot]. Is there some setting in code which i need to do to enable spotless working for me?

Hmm, I ran `mvn clean package -DskipTests` in the root of the project and ended up seeing a bunch of spotless violations. Running `mvn spotless:apply` changed a lot of files in this PR.

vahmed-hamdy commented:
Thanks @19priyadhingra for addressing the comments, could we fix the spotless violations as mentioned above?

19priyadhingra commented:
> Thanks @19priyadhingra for addressing the comments, could we fix the spotless violations as mentioned above?

Yes @vahmed-hamdy , I am trying hard on it. As attached in above screenshots, weirdly, all of the spotless checks are getting skipped for me. For me "mvn clean package -DskipTests" getting succeeded with no spotless warnings. Not sure what mvn/Intelij setting I need to enable in my local build to see all these errors. Trying to take others help on this. 

Thanks!

19priyadhingra commented:
> Thanks @19priyadhingra for addressing the comments, could we fix the spotless violations as mentioned above?

I was able to finally make "spotless" work for my local workspace post downgrading my Java version from java 21 to java 8. Is it like the pre-requisite for this package to use? Should we update that in readMe package. This will help others to save time in debugging such issues later

19priyadhingra commented:
> Thanks for the efforts. Could we remove the ticket link from the PR title, it should be automatically linked using "Autolink"
> 
> We need to add documentation for the new feature as well

Added Documentation commit in this PR

19priyadhingra commented:
> > Would you recommend keeping it normal, String?
> 
> Yes! And if someone needs base64 encoding they can encode it in the `SerializationSchema`.
> 
> > I have a follow up query on spotless violations. How did you run that. Whenever I am trying to do that, it shows build succeeded with no error and I can see "spotless skipped" in build logs[Attached screenshot]. Is there some setting in code which i need to do to enable spotless working for me?
> 
> Hmm, I ran `mvn clean package -DskipTests` in the root of the project and ended up seeing a bunch of spotless violations. Running `mvn spotless:apply` changed a lot of files in this PR.

1. Removed Base64 encoding and fixed spotless issues

hlteoh37 commented on docs/content.zh/docs/connectors/datastream/sqs.md:
```diff
@@ -0,0 +1,134 @@
+# Amazon SQS Sink
+
+The SQS sink writes to [Amazon SQS](https://aws.amazon.com/sqs) using the [AWS v2 SDK for Java](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/home.html). Follow the instructions from the [Amazon SQS Developer Guide](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/welcome.html)
+to setup a SQS.
```
nit: setup an SQS message queue.

19priyadhingra replied:
Updated

hlteoh37 commented on docs/content.zh/docs/connectors/datastream/sqs.md:
````diff
@@ -0,0 +1,134 @@
+```java
+Properties sinkProperties = new Properties();
+// Required
+sinkProperties.put(AWSConfigConstants.AWS_REGION, "eu-west-1");
````
Is there a reason we need to specify this? Can we figure this out via the SQS URL?

19priyadhingra replied:
SQSSinkWriter using [AWSClientUtil -> createAwsAsyncClient](https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSClientUtil.java#L80) method to create SqsAsyncClient client object which internally uses ```AWS_REGION``` props to read client region info. This flow is consistent with creating other AWS sink resources so I have gone through the same. We have to write custom code if we want to derive region info from SQS Url, I didn't see any benefit of doing that

hlteoh37 commented on docs/content.zh/docs/connectors/datastream/sqs.md:
```diff
@@ -0,0 +1,134 @@
+    * Url of the SQS to sink to.
+4. _setFailOnError(boolean failOnError)_
+    * Optional. Default: `false`.
+    * Whether failed requests to write records to SQS are treated as fatal exceptions in the sink.
```
nit: Might be a better description if we say
"causes a Flink job restart"

hlteoh37 commented on docs/content/docs/connectors/datastream/sqs.md:
```diff
@@ -0,0 +1,134 @@
+---
+title: DynamoDB
+weight: 5
+type: docs
+aliases:
+- /zh/dev/connectors/dynamodb.html
```
Same as above - should be SQS instead of DynamoDB

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,122 @@
+            <version>${project.version}</version>
+            <scope>test</scope>
+            <type>test-jar</type>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.typesafe.netty</groupId>
+                    <artifactId>netty-reactive-streams-http</artifactId>
+                </exclusion>
+            </exclusions>
```
Why do we need this exclusion?

19priyadhingra replied:
Unnecessary. Removed in next revision

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,122 @@
+            <version>${project.version}</version>
+            <scope>test</scope>
+            <type>test-jar</type>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.typesafe.netty</groupId>
+                    <artifactId>netty-reactive-streams-http</artifactId>
+                </exclusion>
+            </exclusions>
```
Why do we need this exclusion?

19priyadhingra replied:
Unnecessary. Removed in next revision

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,122 @@
+        </dependency>
+
+        <!-- Other third-party dependencies -->
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
```
Should we scope to `test`?

19priyadhingra replied:
done

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsConfigConstants.java:
```diff
@@ -0,0 +1,31 @@
+@PublicEvolving
+public class SqsConfigConstants {
+
+    public static final String BASE_SQS_USER_AGENT_PREFIX_FORMAT =
```
Could we use `ConfigOption` instead of strings here? Example: https://github.com/apache/flink-connector-aws/blob/38aafb44d3a8200e4ff41d87e0780338f40da258/flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/config/KinesisStreamsSourceConfigConstants.java#L41-L46

19priyadhingra replied:
Sure. updated with ConfigOption

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,257 @@
+                sqsAsyncClient.sendMessageBatch(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
```
The SQS documentation for `sendMessageBatch` is a little bit worrying...
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessageBatch.html
> The result of sending each message is reported individually in the response. Because the batch request can result in a combination of successful and unsuccessful actions, you should check for batch errors even when the call returns an HTTP status code of 200.

Have we verified that if some messages in batch fails, we will get an `err` here?

19priyadhingra replied:
yes, we do validated that and retry failed ones under ```handlePartiallyFailedRequest```

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,257 @@
+        return Optional.empty();
+    }
+
+    @Internal
+    @VisibleForTesting
+    void setSqsAsyncClient(final SqsAsyncClient sqsAsyncClient) {
+        this.sqsAsyncClient = sqsAsyncClient;
+    }
+
+    @Internal
+    @VisibleForTesting
+    void setSdkAsyncHttpClient(final SdkAsyncHttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
```
This is an antipattern, could we instead use a `ClientProvider` as shown, so we can test it without `@VisibleForTesting` instead?
Example: 
https://github.com/apache/flink-connector-aws/blob/38aafb44d3a8200e4ff41d87e0780338f40da258/flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbSinkWriter.java#L139

19priyadhingra replied:
Updated with client provider

hlteoh37 reviewed (changes requested):
Added initial comments

hlteoh37 commented:
We seem to be having quite a few `.` in the class folders. Can we change them to `/` instead?
e.g. `flink-connector-aws/flink-connector-sqs/src/test/java/org.apache.flink/connector.sqs/sink/SqsExceptionClassifiersTest.java

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkElementConverter.java:
```diff
@@ -0,0 +1,105 @@
+        public SqsSinkElementConverter<InputT> build() {
+            Preconditions.checkNotNull(
+                    serializationSchema,
+                    "No SerializationSchema was supplied to the " + "SQS Sink builder.");
```
nit: `+` not needed here

19priyadhingra replied:
Fixed

19priyadhingra commented:
> We seem to be having quite a few `.` in the class folders. Can we change them to `/` instead? e.g. `[flink-connector-aws](https://issues.apache.org/jira/browse/FLINK-connector-aws)/[flink-connector-sqs](https://issues.apache.org/jira/browse/FLINK-connector-sqs)/src/test/java/org.apache.flink/connector.sqs/sink/SqsExceptionClassifiersTest.java

Good catch! Fixed

19priyadhingra commented on docs/content.zh/docs/connectors/datastream/sqs.md:
```diff
@@ -0,0 +1,134 @@
+---
+title: DynamoDB
```
oops! corrected

sap1ens commented:
It'd also be great to mention what permissions are needed for this connector to work. E.g., is `sqs:SendMessage` sufficient?

simulified commented:
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

19priyadhingra commented:
> It'd also be great to mention what permissions are needed for this connector to work. E.g., is `sqs:SendMessage` sufficient?

Good point!, Yes, `sqs:SendMessage`  is sufficient, updated the same in the documentation.

nehahmoe commented:
Hi, when is this planned to be released?

19priyadhingra commented:
> Hi, when is this planned to be released? @19priyadhingra

Unfortunately, right now It is stuck on getting approval, no actions pending at my end at this stage.

duncandee commented:
Waiting patiently for this to be approved.

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,111 @@
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jdk8</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Why do we need a `jdk8` dependency?

19priyadhingra replied:
Not needed. removed

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkElementConverter.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public void open(Sink.InitContext context) {
+        try {
+            serializationSchema.open(
```
can we do `context.asSerializationSchemaInitializationContext()` instead?

19priyadhingra replied:
Curious to know What is the benefit of using this over the existing one?

hlteoh37 replied:
It passes Flink's serialization context over creating a custom one. This is how Flink passes context from upper levels down.

19priyadhingra replied:
Thanks for the detail, Updated with  serializationSchema.open(context.asSerializationSchemaInitializationContext());

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,222 @@
+        CompletableFuture<SendMessageBatchResponse> future =
+                clientProvider.getClient().sendMessageBatch(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
+                    if (err != null) {
+                        handleFullyFailedRequest(err, requestEntries, requestResult);
+                    } else if (response.failed() != null && response.failed().size() > 0) {
+                        handlePartiallyFailedRequest(response, requestEntries, requestResult);
+                    } else {
+                        requestResult.accept(Collections.emptyList());
+                    }
+                });
```
What happens when `handleFullyFailedRequest` throws an `Exception`? Could we handle this?

```
future.whenComplete(...)
         .exceptionally(ex -> getFatalExceptionCons().accept(ex));
```

19priyadhingra replied:
@hlteoh37 My understanding here was that on Exception, whenComplete will put the exception under err object in (response, err) and then it will be handled under if (err != null) block? I can see the same implementation for all the other sinks

hlteoh37 replied:
Yep - my concern is more around if the block within `whenComplete(...)` throws an exception. We've found in other sinks that this leads to a "deadlock" like situation because we don't have `.exceptionally(...)` for the future.

19priyadhingra replied:
Understood, Added the exceptionally block

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,222 @@
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        if (response.failed() != null) {
```
This `if` is redundant as `handlePartiallyFailedRequest` is only called when `response.failed()!=null)`

19priyadhingra replied:
Good catch, Removed

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,222 @@
+            return;
+        }
+
+        if (response.failed() != null) {
```
This is redundant check

19priyadhingra replied:
Removed

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,222 @@
+            for (final BatchResultErrorEntry failedEntry : response.failed()) {
+                final Optional<SendMessageBatchRequestEntry> retryEntry =
+                        getFailedRecord(requestEntries, failedEntry.id());
+                if (retryEntry.isPresent()) {
+                    failedRequestEntries.add(retryEntry.get());
+                }
```
We should log and restart the job by calling `FatalExceptionCons` if we cannot find the entry, as this can cause data loss if we just silently drop it.

19priyadhingra replied:
Agreed, added a log and called FatalExceptionCons to trigger job restart

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsStateSerializer.java:
```diff
@@ -0,0 +1,58 @@
+        final byte[] requestData = new byte[(int) requestSize];
+        in.read(requestData);
+        return SendMessageBatchRequestEntry.builder()
+                .id(UUID.randomUUID().toString())
```
This is a little bit weird that we generate a random UUID each time the stat is serialized. Flink does this transparently - it would be good to keep this in state + ensure the UUID is kept constant for a given request

19priyadhingra replied:
@hlteoh37 We need to generate this request ID at each batch request entry level, that's the SQS request contract format. We can have multiple batch request entry in each batch, hence generated here.

When you suggested generating it at request level, what do you mean by that? Any references for keeping in this state?

hlteoh37 replied:
No that makes sense, but the comment here is more around not re-generating the `UUID` every time we serialize / deserialize the request. Instead, can we include the UUID in the serialization and retrieve it when we deserialize it?

19priyadhingra replied:
Make sense, updated the code accordingly

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/test/java/org/apache/flink/connector/sqs/sink/SqsSinkWriterTest.java:
```diff
@@ -0,0 +1,342 @@
+/** Covers construction, defaults and sanity checking of {@link SqsSinkWriter}. */
+public class SqsSinkWriterTest {
+
+    @Mock private SqsAsyncClient sqsAsyncClient;
+
+    @Mock private Consumer<List<SendMessageBatchRequestEntry>> requestResult;
+
+    @Mock private SdkClientProvider<SqsAsyncClient> asyncClientSdkClientProviderOverride;
+
```
Mockito is not recommended in Flink. Can we instead initialize test classes to mimic expected behavior on AsyncClient side? See example for DDB: https://github.com/apache/flink-connector-aws/blob/36479a98e11c154bd1537e11e8cc51e64a0ebcb8/flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/sink/DynamoDbSinkWriterTest.java#L358

19priyadhingra replied:
Thanks for the reference, removed Mockito dependency

hlteoh37 reviewed (changes requested):
Added some comments. Thanks @19priyadhingra

19priyadhingra commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,111 @@
+    <parent>
+        <artifactId>flink-connector-aws-e2e-tests-parent</artifactId>
+        <groupId>org.apache.flink</groupId>
+        <version>4.3-SNAPSHOT</version>
```
Updated

19priyadhingra commented on flink-connector-aws/flink-connector-sqs/pom.xml:
```diff
@@ -0,0 +1,135 @@
+            <artifactId>jackson-datatype-jsr310</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jdk8</artifactId>
```
Removed

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkBuilder.java:
```diff
@@ -0,0 +1,159 @@
+        return this;
+    }
+
+    /**
+     * Allows the user to specify a serialization schema to serialize each record to persist to SQS.
+     *
+     * @param schema serialization schema to use
+     * @return {@link SqsSinkBuilder} itself
+     */
+    public SqsSinkBuilder<InputT> setSerializationSchema(final SerializationSchema<InputT> schema) {
+        serializationSchema = schema;
+        return this;
+    }
```
Should we instead provide a method to set the Element converter directly? Otherwise users will create SerializationSchema that takes type `InputT` -> `byte[]` -> `String` Rather than directly `InputT` -> `String`

19priyadhingra replied:
I tried to update the same, please let me know if that looks good

hlteoh37 replied:
Looks great - thanks @19priyadhingra !

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java:
```diff
@@ -0,0 +1,235 @@
+        }
+
+        LOG.warn(
+                "SQS Sink failed to write and will retry {} entries to SQS first request was {}",
```
typo: 
```suggestion
                "SQS Sink failed to write and will retry {} entries to SQS. First request was {}",
```

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/main/resources/log4j2.properties:
```diff
@@ -0,0 +1,25 @@
+# limitations under the License.
+################################################################################
+
+rootLogger.level = OFF
```
Can we remove this file?

19priyadhingra replied:
Why do we want to do that? I can see other sinks also have their independent log4j configuration.

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/test/java/org/apache/flink/connector/sqs/sink/SqsSinkElementConverterTest.java:
```diff
@@ -0,0 +1,84 @@
+                SqsSinkElementConverter.<String>builder()
+                        .setSerializationSchema(new SimpleStringSchema())
+                        .build();
+        elementConverter.open(null);
```
This needs to be changed - test fails
```suggestion
        elementConverter.open(new TestSinkInitContext());
```

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/test/java/org/apache/flink/connector/sqs/sink/SqsSinkElementConverterTest.java:
```diff
@@ -0,0 +1,84 @@
+                SqsSinkElementConverter.<String>builder()
+                        .setSerializationSchema(openCheckingStringSchema)
+                        .build();
+        elementConverter.open(null);
```
```suggestion
        elementConverter.open(new TestSinkInitContext());
```

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/test/resources/META-INF/services/org.junit.jupiter.api.extension.Extension:
```diff
@@ -0,0 +1,16 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+org.apache.flink.util.TestLoggerExtension
```
What do we use this for?

19priyadhingra replied:
This got copied from other sink package, doesn't look like it is required, deleting it now

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/pom.xml:
```diff
@@ -0,0 +1,130 @@
+            <artifactId>jackson-datatype-jsr310</artifactId>
+        </dependency>
+
+        <!-- ArchUit test dependencies -->
```
nit `ArchUnit`

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/src/test/java/org/apache/flink/connector/sqs/sink/test/SqsSinkITTest.java:
```diff
@@ -0,0 +1,160 @@
+
+    @Before
+    public void setup() throws Exception {
+        System.setProperty(SdkSystemSetting.CBOR_ENABLED.property(), "false");
```
Do we need this for SQS? I'm aware we need it for KDS, but just checking

19priyadhingra replied:
This need to be validated once i will able to run this test locally

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,105 @@
+
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>copy</id>
+                        <phase>pre-integration-test</phase>
+                        <goals>
+                            <goal>copy</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
```
This should be removed.

hlteoh37 commented on flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,105 @@
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>flink-connector-aws-sqs-e2e-tests</artifactId>
```
This test doesn't seem to work. Please make sure we have all the dependencies.


```
Caused by: java.lang.NoClassDefFoundError: software/amazon/awssdk/services/s3/S3Client
        at org.apache.flink.connector.aws.testutils.AWSServicesTestUtils.createS3Client(AWSServicesTestUtils.java:65)
        at org.apache.flink.connector.aws.testutils.LocalstackContainer$ListBucketObjectsWaitStrategy.list(LocalstackContainer.java:79)
        at org.rnorth.ducttape.ratelimits.RateLimiter.getWhenReady(RateLimiter.java:51)
        at org.apache.flink.connector.aws.testutils.LocalstackContainer$ListBucketObjectsWaitStrategy.lambda$waitUntilReady$0(LocalstackContainer.java:72)
        at org.rnorth.ducttape.unreliables.Unreliables.lambda$retryUntilSuccess$0(Unreliables.java:43)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassNotFoundException: software.amazon.awssdk.services.s3.S3Client
```

I ran using the below command

```
mvn clean verify -Prun-end-to-end-tests -DdistDir=/Users/liangtl/workplace/flink_os/flink-1.19.0/lib/flink-dist-1.19.0.jar -pl flink-connector-aws-e2e-tests/flink-connector-aws-sqs-e2e-tests -am
```

19priyadhingra replied:
Above command didn't work for me but I tried to run the test case by putting the test case file in same sqs-sink/test case folder and test passed successfully there. Please let me know if that still fails at your end.

hlteoh37 commented on docs/content/docs/connectors/datastream/sqs.md:
```diff
@@ -0,0 +1,134 @@
+weight: 5
+type: docs
+aliases:
+   - /zh/dev/connectors/sqs.html
```
I don't think we need `zh` here

19priyadhingra replied:
Removed

hlteoh37 commented:
spotless seems to be failing.

19priyadhingra commented:
> spotless seems to be failing.

oops sorry!! Fixed now

hlteoh37 commented on flink-connector-aws/flink-connector-sqs/src/test/java/org/apache/flink/connector/sqs/sink/SqsSinkBuilderTest.java:
```diff
@@ -25,14 +25,6 @@
-                .withMessageContaining(
-                        "No SerializationSchema was supplied to the SQS Sink builder.");
-    }
-
```
Why did we remove this unit test?

19priyadhingra replied:
because we added a default SerializationSchema if it is not provided by the customer so we never end up throwing this error

hlteoh37 commented:
@19priyadhingra The tests seems to be failing. Can we please take a look?

Also - it would be good if we squash the commits!

19priyadhingra commented:
> @19priyadhingra The tests seems to be failing. Can we please take a look?
> 
> Also - it would be good if we squash the commits!

1. Hi Hong, I tried to deep dive more on the only failed test: elementConverterWillOpenSerializationSchema
failed logs: https://paste.amazon.com/show/dhipriya/1721770192
it complains TestSinkInitContext cant be cast to sink2.Sink$InitContext. I understood the reason why it is failing but not sure about how to fix it. This issue started coming post 1.19.1 where we made change in TestSinkInitContext. One of the easy way could be to remove this test if it is not adding any big value or other fix might require change in flink-connector-base where we have to keep the old TestSinkInitContext? Right now I removed that test, please let me know your thought on this. 
[https://github.com/apache/flink/blob/release-1.19.1/flink-connectors/flink-connect[…]pache/flink/connector/base/sink/writer/TestSinkInitContext.java](https://github.com/apache/flink/blob/release-1.19.1/flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/TestSinkInitContext.java#L51) -> new TestSinkInitContext implements WriterInitContext whereas old TestSinkInitContext (  [https://github.com/apache/flink/blob/release-1.18.1/flink-connectors/flink-connect[…]pache/flink/connector/base/sink/writer/TestSinkInitContext.java](https://github.com/apache/flink/blob/release-1.18.1/flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/TestSinkInitContext.java#L47)) implements Sink.InitContext
Now new one which implements [WriterInitContext](https://github.com/apache/flink/blob/release-1.19.1/flink-core/src/main/java/org/apache/flink/api/connector/sink2/WriterInitContext.java#L35) extends org.apache.flink.api.connector.sink2.InitContext, not the required one org.apache.flink.api.connector.sink2.Sink.InitContext

2. Squashed the commits

19priyadhingra commented:
Integ test are failing because local stack is unable to create sqs, this issue seems like the one mentioned here: https://github.com/aws/aws-sdk-java/issues/3055

But weirdly, the same test case was passed earlier, anybody aware of any changes made in this repo related to local-stack that can cause this?

z3d1k commented:
@19priyadhingra 
This may potentially be related to recent AWS SDK upgrade in #149

19priyadhingra commented:
> #149

Thanks for the info @z3d1k , reverting that change in my local workspace actually fixed the test case

hlteoh37 commented:
> Do you have any suggestion on how can I fix my test case without reverting that PR?

Hi @19priyadhingra, we can either debug the local stack issue, or consider reducing the AWS version for our ITCase against localstack - we need to track the fix with a JIRA though

19priyadhingra commented:
> consider reducing the AWS version for our ITCase against localstack

I am sorry, I didn't understand this part, do you mean replacing 1 with 2? I tried this but build started failing because of conflicting aws-sdk version exists at the same time

1.
<dependency>
    <groupId>software.amazon.awssdk</groupId>
    <artifactId>sqs</artifactId>
</dependency>

2
<dependency>
    <groupId>software.amazon.awssdk</groupId>
    <artifactId>sqs</artifactId>
    <version>2.20.144</version>
</dependency>

19priyadhingra commented:
> > Do you have any suggestion on how can I fix my test case without reverting that PR?
> 
> Hi @19priyadhingra, we can either debug the local stack issue, or consider reducing the AWS version for our ITCase against localstack - we need to track the fix with a JIRA though

Added following in sqs-e2e-sqs pom file

```
    <dependencies>

            <!-- Overridden aws-sdk dependency to older version to temporarily fix 'not able to create sqs localstack error with newer version'-->
            <dependency>
                <groupId>software.amazon.awssdk</groupId>
                <artifactId>bom</artifactId>
                <version>2.20.144</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
```

19priyadhingra commented:
@hlteoh37 All workflows passed in last trigger. is there any other blocker in merging this? Thanks!

VittoriDavide commented:
yes this PR will be super helpful for us as well 👀

hlteoh37 commented:
Hey @19priyadhingra LGTM. Can you please sort out the commit history here? Seems like it's trying to include 14 commits.
Also, it would be good if we created a Tech Debt JIRA to sort out the E2E test AWS version!

19priyadhingra commented:
> Hey @19priyadhingra LGTM. Can you please sort out the commit history here? Seems like it's trying to include 14 commits. Also, it would be good if we created a Tech Debt JIRA to sort out the E2E test AWS version!

Thanks a lot for approving PR. I will certainly create a JIRA soon with all the details on E2E AWS version issue.

Can anybody please help on commit history issue called out here, I am not able to exactly figure out why it is showing 14 commits, instead of just one in my PR. They started showing up when I sync my forked branch from mainline. How can I make it one? I don't think squash can help here as that will include other commit also in my SQS sink commit.

 is it like I have to create new fork and again raise the PR? Thanks!

z3d1k commented:
It looks like you've merged the remote branch into your fork. Instead of doing that you need to rebase your commit on it.
Merging branch rewrites commit ids, so it does not align with upstream branch state.

19priyadhingra commented:
> @19priyadhingra It looks like you've merged the remote branch into your fork. Instead of doing that you need to rebase your commit on it.
> 
> Recreating fork and PR is a bit extreme, here's are steps that can help recover branch (tested on my side, but have not verified content):
> 
> ```
> # get copy of your fork repo, useful to have it separately to verify that no changes were lost
> git clone git@github.com:19priyadhingra/flink-connector-aws.git
> cd flink-connector-aws/
> 
> # add main repo as a remote and get metadata
> git remote add upstream git@github.com:apache/flink-connector-aws.git
> git fetch --all
> 
> # roll back merge commit
> git reset HEAD~1
> 
> # rebase onto upstream branch
> git pull -r upstream main
> ```
> 
> Please let me know if this helped

THANKS A LOT, @z3d1k . This worked :-) Much appreciated for this quick help!!

z3d1k commented:
Great!

z3d1k commented:
@19priyadhingra Please also update the commit and PR name to follow [naming convention](https://flink.apache.org/how-to-contribute/code-style-and-quality-pull-requests/#1-jira-issue-and-naming).
The name should be `[FLINK-35305][Connector/SQS] Amazon SQS Sink Connector`.

19priyadhingra commented:
> @19priyadhingra Please also update the commit and PR name to follow [naming convention](https://flink.apache.org/how-to-contribute/code-style-and-quality-pull-requests/#1-jira-issue-and-naming). The name should be `[FLINK-35305][Connector/SQS] Amazon SQS Sink Connector`.

Sure, updated both

19priyadhingra commented:
@hlteoh37 how can we merge this now?

hlteoh37 reviewed (approved):
LGTM. Thanks @19priyadhingra

nehahmoe commented:
Congratulation @19priyadhingra 
Thanks for the contribution, we have been waiting on this one. 
When will the package be published to maven artifactory?

sap1ens commented:
@19priyadhingra thanks for your work!

VittoriDavide commented:
Thank you @19priyadhingra! 🔥
